### PR TITLE
Make sampler less blocking

### PR DESCRIPTION
**Pull request by Yard1**
This PR reduces sampling overhead by rearranging operations to ensure GPU can run uninterrupted for longer. This is achieved by:
1. Preparing sampling tensors like temperature ahead of time and in a separate CUDA stream
2. Using a custom multinomial sampling implementation that doesn't require a CPU<->GPU sync
3. Delaying the CPU<->GPU sync for longer.

Pytorch profiler output for current master:
Screenshot 2023-12-01 at 1 10 13 PM

Pytorch profiler output for this PR:
Screenshot 2023-12-01 at 1 10 33 PM

The impact of the changes is clearly visible by making the gaps between kernel launches smaller. This allows us to save ~2ms per iteration.

There are no changes to the outputs - this is strictly a refactor.

### CODE REVIEW

**Comment by Yard1** [comment]
The changes to the top_p/top_k sampler made in https://github.com/vllm-project/vllm/pull/1885 should override the changes in this PR.

File: vllm/model_executor/sampling_metadata.py
Comment (zhuohan123) [inline vllm/model_executor/sampling_metadata.py:165]: There was an issue of `pin_memory` in WSL as in below:

https://github.com/vllm-project/vllm/blob/e19a64c7eff2085790dbf71851208fa2dd31ca4d/vllm/worker/cache_engine.py#L92-L110

Can you modify the code here accordingly?

File: vllm/model_executor/sampling_metadata.py
Comment (zhuohan123) [inline vllm/model_executor/sampling_metadata.py:157]: Curious how big an overhead the small tensor copies can cause? I am worried the code below can be a little bit too complex compared to the benefits it brings.

File: vllm/model_executor/layers/sampler.py
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:30]: Is there any specific reason to make this a global property instead of having a stream per-sampler?

File: vllm/model_executor/layers/sampler.py
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:228]: Let's keep the definition of 
```python
top_p_mask = probs_sum > p.unsqueeze(dim=1)
```
and
```python
top_k_mask = top_k_mask >= k.unsqueeze(dim=1)
``` 
above and let
```python
mask = top_p_mask | top_k_mask
```
This should not impact performance but make the code more clear

File: vllm/model_executor/sampling_metadata.py
Comment (zhuohan123) [inline vllm/model_executor/sampling_metadata.py:107]: Add a comment:
```suggestion
            if (i < sampling_metadata.num_prompts
                    and sampling_params.prompt_logprobs is not None):
                # For tokens in the prompt that we only need to get their logprobs
                prompt_len = sampling_metadata.prompt_lens[i]
```

File: vllm/worker/model_runner.py
Comment (zhuohan123) [inline vllm/worker/model_runner.py:29]: When will `model_config` be None?

**Review by zhuohan123 — COMMENTED** [review]
Thank you for the great PR! I left some comments, mainly on the code style.

**Comment by WoosukKwon** [comment]
@zhuohan123 Thanks for the review! Just to be safe, please wait for the release before merging the PR.

File: vllm/model_executor/sampling_metadata.py
Comment (Yard1) [reply vllm/model_executor/sampling_metadata.py:157]: I can break it up here, it doesn't matter much in this particular case. Overall I found it to be beneficial, as it reduces the amount of kernel launches and python overhead noticeably.

File: vllm/model_executor/layers/sampler.py
Comment (Yard1) [reply vllm/model_executor/layers/sampler.py:228]: Sure. This will be superseded by the other PR regardless

File: vllm/model_executor/sampling_metadata.py
Comment (Yard1) [reply vllm/model_executor/sampling_metadata.py]: Broke it up - it doubles the python time for the function but the time is still sub 1ms and it happens asynchronously so it shouldn't be a big issue.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the hard work!

**Pull request merged**
